src/coreComponents/constitutive/solid/SolidBase.hpp

```diff
@@ -202,63 +588,79 @@ class SolidBase : public constitutive::ConstitutiveBase
   virtual void allocateConstitutiveData( dataRepository::Group * const parent,
                                          localIndex const numConstitutivePointsPerParentIndex ) override;
 
-
+  virtual void saveConvergedState();
+  
   struct viewKeyStruct : public ConstitutiveBase::viewKeyStruct
   {
-    static constexpr auto defaultDensityString  = "defaultDensity";
-    static constexpr auto densityString  = "density";
     static constexpr auto stressString = "stress";
+    static constexpr auto oldStressString = "oldStress";
+    static constexpr auto densityString  = "density";
+    static constexpr auto defaultDensityString  = "defaultDensity";
   };
 
+
+  /// Number of elements storing solid data
+  localIndex numElem() const
+  {
+    return m_newStress.size( 0 );
+  }
+  
+  /// Number of quadrature points per element storing solid data
+  localIndex numQuad() const
+  {
+    return m_newStress.size( 1 );
+  }
+  
   /**
    * @name Accessors
    */
   ///@{
 
-  /**
-   * Getter for default density
-   * @return The default density
-   */
-  real64 getDefaultDensity() const
+  /// Non-const/mutable accessor for stress
+  arrayView3d< real64, solid::STRESS_USD > const getStress()
   {
-    return m_defaultDensity;
+    return m_newStress;
   }
 
-  /**
-   * Setter for default density
-   * @param density The value to set m_defaultDensity equal to.
-   */
-  void setDefaultDensity( real64 const density )
+  /// Const/non-mutable accessor for stress
+  arrayView3d< real64 const, solid::STRESS_USD > const getStress() const
   {
-    m_defaultDensity = density;
+    return m_newStress;
   }
 
   /// Non-const/Mutable accessor for density.
-  arrayView2d< real64 > getDensity() { return m_density; }
+  arrayView2d< real64 > const getDensity()
+  {
+    return m_density;
+  }
 
   /// Const/non-mutable accessor for density
-  arrayView2d< real64 const > getDensity() const { return m_density; }
-
-  /// Non-const/mutable accessor for stress
-  arrayView3d< real64, solid::STRESS_USD > getStress() { return m_stress; }
-
-  /// Const/non-mutable accessor for stress
-  arrayView3d< real64 const, solid::STRESS_USD > getStress() const { return m_stress; }
-
+  arrayView2d< real64 const > const getDensity() const
+  {
+    return m_density;
+  }
+  
   ///@}
 
 protected:
-  virtual void PostProcessInput() override;
 
-  /// The default density for new allocations.
-  real64 m_defaultDensity = 0;
+  virtual void PostProcessInput() override;
 
+  /// The current stress at a quadrature point (i.e. at timestep n, global newton iteration k)
+  array3d< real64, solid::STRESS_PERMUTATION > m_newStress;
+  
+  /// The previous stress at a quadrature point (i.e. at timestep (n-1))
+  array3d< real64, solid::STRESS_PERMUTATION > m_oldStress;
```

Comment:
With two stress quantities in all cases, we are doubling the amount of data that needs to be moved around for an update.

Reply:
- For plasticity models, we need both a new state and an old state.  There is no way around it.  It also makes it convenient to wind back in time if a timestep fails, and to call the elasticity model for the trial plastic state.  
- For elasticity, we have the "NoState" model where m_newStress and m_oldStress are not allocated at all.  Can't we just call this version if we want a memory optimal model?  This won't account for a given initial stress field though. We could allocate m_newStress but not m_oldStress, and store the initial state in m_newStress.